pkg/nexus/nexus.go

```diff
@@ -0,0 +1,298 @@
+type Upload struct {
+	baseURL   string
+	version   string
+	Username  string
+	Password  string
+	artifacts []ArtifactDescription
+	Logger    *logrus.Entry
+}
+
+// Uploader provides an interface to the nexus upload for configuring the target Nexus Repository and
+// adding artifacts.
+type Uploader interface {
+	SetBaseURL(nexusURL, nexusVersion, repository, groupID string) error
+	SetArtifactsVersion(version string) error
+	AddArtifact(artifact ArtifactDescription) error
+	AddArtifactsFromJSON(json string) error
+	GetArtifacts() []ArtifactDescription
+	UploadArtifacts() error
+}
+
+func (nexusUpload *Upload) initLogger() {
+	if nexusUpload.Logger == nil {
+		nexusUpload.Logger = log.Entry().WithField("package", "SAP/jenkins-library/pkg/nexusUpload")
+	}
+}
+
+// SetBaseURL constructs the base URL for all uploaded artifacts. No parameter can be empty.
+func (nexusUpload *Upload) SetBaseURL(nexusURL, nexusVersion, repository, groupID string) error {
+	if nexusURL == "" {
+		return errors.New("nexusURL must not be empty")
+	}
+	if nexusVersion != "nexus2" && nexusVersion != "nexus3" {
+		return errors.New("nexusVersion must be one of 'nexus2' or 'nexus3'")
+	}
+	if repository == "" {
+		return errors.New("repository must not be empty")
+	}
+	if groupID == "" {
+		return errors.New("groupID must not be empty")
+	}
+	baseURL, err := getBaseURL(nexusURL, nexusVersion, repository, groupID)
+	if err != nil {
+		return err
+	}
+	nexusUpload.baseURL = baseURL
+	return nil
+}
+
+// SetArtifactsVersion sets the common version for all uploaded artifacts. The version is external to
+// the artifact descriptions so that it is consistent for all of them.
+func (nexusUpload *Upload) SetArtifactsVersion(version string) error {
+	if version == "" {
+		return errors.New("version must not be empty")
+	}
+	nexusUpload.version = version
+	return nil
+}
+
+// AddArtifact adds a single artifact to be uploaded later via UploadArtifacts(). If an identical artifact
+// description is already contained in the Upload, the function does nothing and returns no error.
+func (nexusUpload *Upload) AddArtifact(artifact ArtifactDescription) error {
+	err := validateArtifact(artifact)
+	if err != nil {
+		return err
+	}
+	if nexusUpload.containsArtifact(artifact) {
+		log.Entry().Infof("Nexus Upload already contains artifact %v\n", artifact)
+		return nil
+	}
+	nexusUpload.artifacts = append(nexusUpload.artifacts, artifact)
+	return nil
+}
+
+func validateArtifact(artifact ArtifactDescription) error {
+	if artifact.File == "" || artifact.ID == "" || artifact.Type == "" {
+		return fmt.Errorf("Artifact.File (%v), ID (%v) or Type (%v) is empty",
+			artifact.File, artifact.ID, artifact.Type)
+	}
+	return nil
+}
+
+// AddArtifactsFromJSON parses the provided JSON string into an array of ArtifactDescriptions and adds each of
+// them via AddArtifact().
+func (nexusUpload *Upload) AddArtifactsFromJSON(json string) error {
```

Comment:
as discussed: the "user interface" to pass artifacts as json suits more to the cmd which will be available via command line